src/blk/BlockDevice.cc

```diff
@@ -90,67 +86,63 @@ void IOContext::release_running_aios()
 #endif
 }
 
-BlockDevice *BlockDevice::create(CephContext* cct, const string& path,
-				 aio_callback_t cb, void *cbpriv, aio_callback_t d_cb, void *d_cbpriv)
-{
-  string type = "kernel";
-  char buf[PATH_MAX + 1];
-  int r = ::readlink(path.c_str(), buf, sizeof(buf) - 1);
-  if (r >= 0) {
-    buf[r] = '\0';
-    char *bname = ::basename(buf);
-    if (strncmp(bname, SPDK_PREFIX, sizeof(SPDK_PREFIX)-1) == 0)
-      type = "ust-nvme";
-  }
+BlockDevice::block_device_t
+BlockDevice::detect_device_type(const std::string& path) {
+  block_device_t device_type = dev_aio;
 
+#if defined(HAVE_SPDK)
+  if (NVMEDevice::support(path)) {
+    device_type = dev_spdk;
+  } else
+#endif
 #if defined(HAVE_BLUESTORE_PMEM)
-  if (type == "kernel") {
-    int is_pmem = 0;
-    size_t map_len = 0;
-    void *addr = pmem_map_file(path.c_str(), 0, PMEM_FILE_EXCL, O_RDONLY, &map_len, &is_pmem);
-    if (addr != NULL) {
-      if (is_pmem)
-	type = "pmem";
-      else
-	dout(1) << path.c_str() << " isn't pmem file" << dendl;
-      pmem_unmap(addr, map_len);
-    } else {
-      dout(1) << "pmem_map_file:" << path.c_str() << " failed." << pmem_errormsg() << dendl;
-    }
+  if (device_type == dev_aio && PMEMDevice::support(path)) {
+    device_type = dev_pmem;
+  } else
+#endif
+#if (defined(HAVE_LIBAIO) || defined(HAVE_POSIXAIO)) && defined(HAVE_LIZBC)
+  if (device_type == dev_aio && HMSMRDevice::support(path)) {
+    device_type = dev_hmsmr;
   }
 #endif
+  ;
 
-  dout(1) << __func__ << " path " << path << " type " << type << dendl;
+  return device_type;
+}
 
-#if defined(HAVE_BLUESTORE_PMEM)
-  if (type == "pmem") {
-    return new PMEMDevice(cct, cb, cbpriv);
-  }
-#endif
+BlockDevice* BlockDevice::create_with_type(block_device_t device_type,
+  CephContext* cct, const std::string& path, aio_callback_t cb,
+  void *cbpriv, aio_callback_t d_cb, void *d_cbpriv) {
```

Comment:
could you move the opening "{" to a new line?

Reply:
done